app/controllers/UnassignedClientController.scala

```diff
@@ -146,6 +145,73 @@ class UnassignedClientController @Inject()(
     }
   }
 
+  def showConfirmRemoveClient(clientId: String): Action[AnyContent] = Action.async { implicit request =>
+    isAuthorisedAgent { arn =>
+      isOptedInComplete(arn) { _ =>
+        withSessionItem(SELECTED_CLIENTS) { selectedClients =>
+          selectedClients.getOrElse(Seq.empty).find(_.id == clientId)
+            .fold {
+              Redirect(controller.showUnassignedClients(None)).toFuture
+            } { client =>
+              sessionCacheService
+                .put(CLIENT_TO_REMOVE, client)
+                .flatMap(_ =>
+                  Ok(
+                    confirm_remove_client(
+                      YesNoForm.form(),
+                      "",
+                      client,
+                      backLink = controller.showSelectedUnassignedClients(None, None),
+                      formAction = controller.submitConfirmRemoveClient
+                  )
+                  ).toFuture
+                )
+            }
+        }
+      }
+    }
+  }
+
+  def submitConfirmRemoveClient: Action[AnyContent] = Action.async { implicit request =>
+    isAuthorisedAgent { arn =>
+      isOptedInComplete(arn) { _ =>
+      withSessionItem[DisplayClient](CLIENT_TO_REMOVE) { maybeClient =>
+        withSessionItem[Seq[DisplayClient]](SELECTED_CLIENTS) { maybeSelectedClients =>
+          if (maybeClient.isEmpty || maybeSelectedClients.isEmpty) {
+            Redirect(controller.showUnassignedClients(None)).toFuture
+          }
+          else {
+            YesNoForm
+              .form("group.client.remove.error")
+              .bindFromRequest
+              .fold(
+                formWithErrors => {
+                  Ok(
+                    confirm_remove_client(
+                      formWithErrors,
+                      "",
+                      maybeClient.get,
+                      backLink = controller.showSelectedUnassignedClients(None, None),
+                      formAction = controller.submitConfirmRemoveClient
+                    )
+                  ).toFuture
+                }, (yes: Boolean) => {
+                  if (yes) {
+                    val clientsMinusRemoved = maybeSelectedClients.get.filterNot(_ == maybeClient.get)
+                    sessionCacheService
+                      .put(SELECTED_CLIENTS, clientsMinusRemoved)
+                      .map(_ => Redirect(controller.showSelectedUnassignedClients(None, None)))
+                  }
+                  else Redirect(controller.showSelectedUnassignedClients(None, None)).toFuture
+                }
+              )
+          }
+        }
+      }
+    }
+    }
+  }
+
```

Comment:
Can you adapt this to accept an optional clientId like in `app/controllers/CreateGroupSelectClientsController.scala` 
def showConfirmRemoveClient(clientId: Option[String]):

So that you don't end up with the same bug (related to triggering form errors or switching language losing the clientId from the url)

Reply:
Ok. done that now.